### PR TITLE
In case of MPS device also copy batch to CPU

### PR DESCRIPTION
**Pull request by hyenal**
# What does this PR do?

This PR fixes an issue when using MPS device and composer's classifier. When computing metrics, `outputs` is copied to the cpu but not `batch` which creates some downstream error.

# What issue(s) does this change relate to?

Fixes https://github.com/mosaicml/composer/issues/3094

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -3327,12 +3337,14 @@ def _eval_loop(
                                                 outputs.append(v)
                                     else:
                                         outputs = self.state.outputs.cpu()
+                                    batch = DeviceCPU().batch_to_device(batch)
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:3340]:
```suggestion
                                    batch = DeviceCPU().batch_to_device(self.state.batch)
```

**Review by mvpatel2000 — COMMENTED** [review]
Thanks for the PR!

I think there's a minor error -- once fixed looks good to me!

File: composer/trainer/trainer.py (hunk 1 above)
Comment (hyenal) [reply composer/trainer/trainer.py]: Thank you this is now integrated in this PR

**Comment by mvpatel2000** [comment]
Hm... labels can be a complex type so directly calling to seems to fail tests

**Comment by hyenal** [comment]
To ensure compatibility with `HuggingFaceModel` I had to create a new method to shift labels outside of `eval_forward`. 

I hope this is not a breaking change, it felt a bit more natural to modify the labels outside of this method as well but there may be reasons for doing it there that I am unaware of.

In addition `update_metric` will be slightly less efficient although the shift label operation should be very cheap

File: composer/models/huggingface.py [hunk 2]
```diff
@@ -544,26 +543,19 @@ def eval_forward(self, batch, outputs: Optional[Any] = None):
                 ]
 
         if self.use_logits or batch.get('mode', None) == 'icl_task':
-            # pop labels first to avoid computing loss
```
[the remark this answers is not in the document]
Comment (hyenal) [reply composer/models/huggingface.py]: Instead of `outputs` I took `metric` instead but overall this should be the same. Due to the variable nature of `labels` I used `batch_to_device` to handle multiple cases at once

**Review by dakinggg — APPROVED** [review]
